Add DoesNotExist label operator.

bgrant0607 opened the pull request:
@mikedanese @davidopp PTAL. This is just to demonstrate how DoesNotExist could be implemented as discussed in #15350.

However, I found when implementing it that it should already be possible to express nonexistence, using notin with an empty value set.

If we want to go forward with this, more tests are needed.

bgrant0607 committed:
Add DoesNotExist label operator.

davidopp commented:
I generally favor parsimony so I am tempted to say we should go with "notin \<empty\>" but I think what you did here will be easier for users, so I think we should go with the approach in this PR.

bgrant0607 commented:
I have yet to confirm whether `k notin ()` translates to an empty list or a list containing an empty string.

bgrant0607 commented:
Also note that it appears that multiple requirements per key are currently possible.

On the plus side, that means one could specify `k,k notin (v)` to ensure that `k` exists. 

On the minus side, I haven't had time to think about the full implications of that.

For better or worse, the documentation leaves lots of wiggle room.

davidopp commented:
I think multiple requirements per key would be confusing and I'd be in favor of prohibiting it in validation. When I looked at how some of the other open-source systems handle this stuff, they all seemed to restrict to a single constraint per key.

bgrant0607 commented:
notin doesn't accept an empty values list:
```
func TestRequirementConstructor(t *testing.T) {
        requirementConstructorTests := []struct {
                Key     string
                Op      Operator
                Vals    sets.String
                Success bool
        }{
...
                {"x", NotInOperator, sets.NewString(), false},
```

`()` parses to a list containing the empty string:
```
               {&LabelSelector{
                        getRequirement("x", NotInOperator, sets.NewString("abc"), t),
                        getRequirement("y", InOperator, sets.NewString("jkl", "mno"), t),
                        getRequirement("z", NotInOperator, sets.NewString(""), t)},
                        "x notin (abc),y in (jkl,mno),z notin ()", true},
```

bgrant0607 commented:
Added more tests. Will update documentation next.

bgrant0607 commented:
Updated labels.md. PTAL.

davidopp commented on docs/user-guide/labels.md:
```diff
@@ -94,23 +112,25 @@ tier != frontend
 
 The former selects all resources with key equal to `environment` and value equal to `production`.
-The latter selects all resources with key equal to `tier` and value distinct from `frontend`.
+The latter selects all resources with key equal to `tier` and value distinct from `frontend` and all resources with no labels with the `tier` key.
```
add a comma before "and" even though it is grammatically incorrect, to make it easier to parse this sentence

davidopp commented on docs/user-guide/labels.md:
```diff
@@ -94,23 +112,25 @@ tier != frontend
 
 The first example selects all resources with key equal to `environment` and value equal to `production` or `qa`.
-The second example selects all resources with key equal to `tier` and values other than `frontend` and `backend`.
+The second example selects all resources with key equal to `tier` and values other than `frontend` and `backend` and all resources with no labels with the `tier` key.
```
same thing, comma before "and all"

davidopp commented on docs/user-guide/labels.md:
```diff
@@ -126,38 +148,39 @@ LIST and WATCH operations may specify label selectors to filter the sets of obje
+#### Service and ReplicationController
 
-The set of pods that a `service` targets is defined with a label selector. Similarly, the population of pods that a `replicationcontroller` is monitoring is also defined with a label selector. For management convenience and consistency, `services` and `replicationcontrollers` may themselves have labels and would generally carry the labels their corresponding pods have in common. Labels selectors  ...
+The set of pods that a `service` targets is defined with a label selector. Similarly, the population of pods that a `replicationcontroller` should monitor is also defined with a label selector. For management convenience and consistency, `services` and `replicationcontrollers` may themselves have labels and would generally carry the labels their corresponding pods have in common. Labels selectors ...
```
maybe s/monitor/manage/

"would generally carry" is pretty ambiguous; you could make it clearer that this is describing a "best practice" and not something the system does automatically

davidopp commented on docs/user-guide/labels.md:
````diff
@@ -165,26 +188,25 @@ or
+    - {key: environment, operator: NotIn, values: [dev]}
+```
 
+`matchLabels` is a map of `{key,value}` pairs. A single `{key,value}` in the `matchLabels` map is equivalent to an element of `matchExpressions`, whose `key` field is "key", the `operator` is "In", and the `values` array contains only "value". The requirements are ANDed. `matchExpressions` is a list of pod selector requirements. The requirements are ANDed. Valid operators include In, NotIn, Exist ...
````
Say it's a map not a multi-map, assuming we enforce that invariant in validation (which I think we should).

mikedanese replied:
We don't currently enforce it.

davidopp replied:
Maybe it's redundant but should you say that the result of matchLabels and matchExpressions are ANDed? (As written, it's clear that each of them is internally ANDed, but not that the final result is obtained by ANDing the other two results.)

bgrant0607 replied:
Reworded.

HaiyangDING commented:
CC @HaiyangDING

mikedanese commented on pkg/labels/selector.go:
```diff
@@ -106,6 +107,7 @@ func NewRequirement(key string, op Operator, vals sets.String) (*Requirement, er
 			return nil, fmt.Errorf("exact match compatibility requires one single value")
 		}
 	case ExistsOperator:
+	case DoesNotExistOperator:
```
Why don't DoesNotExistOperator, ExistsOperator cases assert that len(vals) == 0

bgrant0607 replied:
Oops. Missed this one. Will do.

bgrant0607 replied:
Done.

mikedanese commented on pkg/labels/selector.go:
```diff
@@ -513,7 +522,12 @@ func (p *Parser) parseRequirement() (*Requirement, error) {
 // in case of no operator 'in, notin, ==, =, !=' are found
```
comment should include !

mikedanese commented on pkg/labels/selector.go:
```diff
@@ -533,6 +548,7 @@ func (p *Parser) parseKeyAndInferOperator() (string, Operator, error) {
 func (p *Parser) parseOperator() (op Operator, err error) {
 	tok, lit := p.consume(KeyAndOperator)
 	switch tok {
+	// DoesNotExistToken shouldn't be here because it's a unary operator, not a binary operator
```
And ExistsToken?

bgrant0607 replied:
There is no ExistsToken.

mikedanese commented on pkg/labels/selector.go:
```diff
@@ -457,7 +466,7 @@ func (p *Parser) parse() ([]Requirement, error) {
 		tok, lit := p.lookahead(Values)
 		switch tok {
-		case IdentifierToken:
+		case IdentifierToken, DoesNotExistToken:
```
Error message on 490 needs update. Can't comment there.

bgrant0607 replied:
Done.

mikedanese commented:
Code looks correct. Thanks for the PR!

bgrant0607 commented:
Addressed feedback, except for the map vs. multi-map issue.

Restricting to a single requirement per key would be a breaking API change, unless the change were only made for Job and not for LIST/WATCH filters.

My other concern about that restriction is that it would preclude some useful combinations. We could do smarter validation, such as preventing simultaneous specification of In and NotIn for the same key. I'd prefer that be done in a separate PR, though, since I have other fires burning.

mikedanese commented:
LGTM, assigning to @davidopp

bgrant0607 commented:
I'll squash after final lgtm.

bgrant0607 commented:
All green.

davidopp commented on docs/user-guide/labels.md:
```diff
@@ -94,23 +112,25 @@ tier != frontend
 ### _Set-based_ requirement
 
-_Set-based_ label requirements allow filtering keys according to a set of values. Matching objects must have all of the specified labels (i.e. all keys and at least one of the values specified for each key). Three kinds of operators are supported: `in`,`notin` and exists (only the key identifier). For example:
+_Set-based_ label requirements allow filtering keys according to a set of values. Three kinds of operators are supported: `in`,`notin` and exists (only the key identifier). For example:
```
exists should be in backticks

bgrant0607 replied:
Exists isn't a keyword.

bgrant0607 commented:
Addressed comments and squashed, since it seems to be converging.

bgrant0607 commented:
@k8s-bot test this

bgrant0607 commented:
Shippable timeout was a flake.

davidopp commented:
LGTM
